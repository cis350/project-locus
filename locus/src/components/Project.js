/* eslint-disable no-alert */
import React, { useState, useEffect } from 'react';
import { useNavigate } from 'react-router-dom';
import {
  Button,
  Row,
  Col,
  Modal,
} from 'react-bootstrap';
import { createProject, getAllProjects } from '../modules/api';

const Project = function ProjectComponent({
  userId,
  email,
  club,
  role,
  setCurrProject,
  setCurrProjectWithoutSpace,
}) {
  const [allProjects, setAllProjects] = useState([]);
  const [modalClick, setModalClick] = useState(false);
  const [projectName, setProjectName] = useState('');
  const [projectAdded, setProjectAdded] = useState(false);

  const navigate = useNavigate();

  useEffect(() => {
    getAllProjects(club).then((res) => {
      if (res.status === 200) {
        setAllProjects(res.jsonContent.result);
      }
    });
  }, [projectAdded]);

  const updateProjectName = ((e) => {
    setProjectName(e.target.value);
  });

  const openModal = () => {
    setModalClick(true);
  };

  const closeModal = () => {
    setModalClick(false);
  };

  const addProject = (() => {
    createProject(club, projectName, email, email).then((res) => {
      if (res.status === 201) {
        setProjectAdded(!projectAdded);
      } else {
        alert('Project already exists');
      }
    });
  });

  const goToManageProject = ((noSpaceProject, project) => {
    setCurrProjectWithoutSpace(noSpaceProject);
    setCurrProject(project);
    navigate(`/projects/manage-projects/${noSpaceProject}/${userId}`);
  });

  const modal = (() => (
    <Modal show={modalClick} onHide={() => closeModal()}>
      <Modal.Header closeButton>
        <Modal.Title>
          Create Project for
          {' '}
          {club}
        </Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <input className="form-control" type="text" placeholder="Type in the project name" onChange={(e) => updateProjectName(e)} />
      </Modal.Body>
      <Modal.Footer>
        <Button className="btn btn-secondary mt-2" onClick={() => closeModal()}>
          Close
        </Button>
        <Button className="btn btn-primary mt-2" onClick={() => { addProject(); closeModal(); }}>
          Create
        </Button>
      </Modal.Footer>
    </Modal>
  ));

  const displayProjects = (() => (
    allProjects.map((data) => {
      if (data.members.includes(email)) {
        return (
          <div className="club-item" key={data.projectName}>
            {/* referenced https://www.delftstack.com/howto/javascript/javascript-remove-spaces */}
            <Button className="club-button" onClick={() => goToManageProject(data.projectName.replace(/\s+/g, ''), data.projectName)}>
              <Row>
                <Col className="d-flex justify-content-center">
                  {data.projectName}
                </Col>
              </Row>
            </Button>
          </div>
        );
      }
      return null;
    })
  ));

  const createNewProjectButton = (() => (
    <div className="club-item">
      {/* referenced https://www.delftstack.com/howto/javascript/javascript-remove-spaces */}
      <Button className="club-button" onClick={() => openModal()}>
        <Row>
          <Col className="d-flex justify-content-center">
            Create new project
          </Col>
        </Row>
      </Button>
    </div>
  ));

  const msgNoProjects = (() => (
    <h4 className="text-center">
      {club}
      {' '}
      has no projects.
    </h4>
  ));

  const hasAtLeastOneProjectAssigned = (() => {
    for (let i = 0; i < allProjects.length; i += 1) {
      if (allProjects[i].members.includes(email)) {
        return true;
      }
    }
    return false;
  });

  const msgNoProjectsAssigned = (() => (
    <h3 className="text-center">
      You do not have any projects assigned for
      {' '}
      {club}
      !
    </h3>
  ));

  return (
    <div className="container" style={{ position: 'relative', padding: '20px' }}>
      {allProjects.length === 0 && msgNoProjects()}
      {(role === 'master' || role === 'admin') && createNewProjectButton()}
      {allProjects.length !== 0 && displayProjects()}
      {!hasAtLeastOneProjectAssigned() && msgNoProjectsAssigned()}
      {modalClick && (modal())}
    </div>
  );
};

export default Project;
