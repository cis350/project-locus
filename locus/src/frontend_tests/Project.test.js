/**
* @jest-environment jsdom
*/
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import Project from '../components/Project';

const axios = require('axios');
const MockAdapter = require('axios-mock-adapter');

// mock useNavigate: https://github.com/remix-run/react-router/issues/7811
const mockedNavigator = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigator,
}));

let mock;
beforeAll(() => {
  mock = new MockAdapter(axios);
});
afterEach(() => {
  mock.reset();
});

test('project snapshot test', () => {
  const component = renderer.create(<Project />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
