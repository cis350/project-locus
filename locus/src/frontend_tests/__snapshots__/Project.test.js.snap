// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project snapshot test 1`] = `
<div
  className="container"
  style={
    Object {
      "padding": "20px",
      "position": "relative",
    }
  }
>
  <h4
    className="text-center"
  >
     
    has no projects.
  </h4>
</div>
`;
